nixos/modules/services/continuous-integration/buildkite-agent.nix

```diff
@@ -72,27 +72,20 @@ in
     environment.systemPackages = [ pkgs.buildkite-agent ];
 
     systemd.services.buildkite-agent =
-      let copyOrEcho       = x: target: perms:
-                             (if isPath x
-                              then "cp -f ${x} ${target}; "
-                              else "echo '${x}' > ${target}; ")
-                             + "${pkgs.coreutils}/bin/chmod ${toString perms} ${target}; ";
-          catOrLiteral     = x:
-                             (if isPath x
-                              then "$(cat ${toString x})"
-                              else "${x}");
+      let copy = x: target: perms:
+                 "cp -f ${x} ${target}; ${pkgs.coreutils}/bin/chmod ${toString perms} ${target}; ";
       in
       { description = "Buildkite Agent";
         wantedBy = [ "multi-user.target" ];
         after = [ "network.target" ];
         environment.HOME = "/var/lib/buildkite-agent";
         preStart = ''
             ${pkgs.coreutils}/bin/mkdir -m 0700 -p /var/lib/buildkite-agent/.ssh
-            ${copyOrEcho cfg.openssh.privateKey "/var/lib/buildkite-agent/.ssh/id_rsa"     600}
-            ${copyOrEcho cfg.openssh.publicKey  "/var/lib/buildkite-agent/.ssh/id_rsa.pub" 600}
+            ${copy cfg.openssh.privateKeyPath "/var/lib/buildkite-agent/.ssh/id_rsa"     600}
+            ${copy cfg.openssh.publicKeyPath  "/var/lib/buildkite-agent/.ssh/id_rsa.pub" 600}
 
             cat > "/var/lib/buildkite-agent/buildkite-agent.cfg" <<EOF
-            token="${catOrLiteral cfg.token}"
+            token="$(cat ${toString cfg.tokenPath})"
```

Comment:
One big caveat here is that tokenPath needs to be on the machine running this build, so not something NixOS itself managed (but NixOps does).

I think we should clarify this in option descriptions.